docs/coreneuron/compatibility.rst

```diff
@@ -44,17 +44,113 @@ Note that models using the following features cannot presently be simulated with
+  Cai = _ion_Cai;
+  ENDVERBATIM
+
+* ``PROCEDURE`` blocks shouldn't return any value.
+* Make sure that all the C/C++ functions called in ``VERBATIM`` blocks are defined in a C/C++ header file which is included in the MOD file.
+
 THREADSAFE
 **********
-CoreNEURON's acceleration is made possible through it's ability not only to run the simulations in parallel but also by being able to take advantage of the vectorization support of CPUs and execution on GPUs. To be able to execute a model in parallel using multiple threads the mod file developer should make sure that the mod file is thread safe. You can find more information about ``THREADSAFE`` mod files `here <https://neuron.yale.edu/neuron/docs/multithread-parallelization>`_.
+CoreNEURON's acceleration is made possible through it's ability not only to run the simulations in parallel but also by being able to take advantage of the vectorization support of CPUs and execution on GPUs. To be able to execute a model in parallel using multiple threads the MOD file developer should make sure that the MOD file is thread safe. You can find more information about ``THREADSAFE`` MOD files `here <https://neuron.yale.edu/neuron/docs/multithread-parallelization>`_.
+
+By making the MOD file thread safe the initialization and current and state update for all the sections that include this mechanism can be done in parallel using vectorized CPU instructions and can be executed in GPU safely.
+
+Some useful instructions to make a MOD file thread safe are:
+
+* Make sure that all the variables that are written in the MOD file and are defined as ``GLOBAL`` are now defined as ``RANGE``. More information about ``GLOBAL`` and ``RANGE`` variables can be found `here <https://nrn.readthedocs.io/en/latest/hoc/modelspec/programmatic/mechanisms/nmodl2.html>`_.
+* In case there are any ``VERBATIM`` blocks those need to be threadsafe and there must be a ``THREADSAFE`` keyword in the ``NEURON`` block to declare that the ``VERBATIM`` blocks are thread safe. Also make sure that the ``NEURON`` block is defined before any ``VERBATIM`` block in the MOD file
+
+For example here is a showcase MOD file that defines `minf` as a ``GLOBAL`` variable which gets updated during the simulation and also defines a ``VERBATIM`` block before the ``NEURON`` block:
+
+.. code-block:: c++
+
+  VERBATIM
+  #include <stdlib.h>
+  ENDVERBATIM
+
+  NEURON {
+    SUFFIX test
+    GLOBAL minf
+  }
+
+  ASSIGNED {
+    v            (mV)
+    minf
+  }
+
+  STATE {
+    m
+  }
+
+  BREAKPOINT {
+    SOLVE states METHOD cnexp
+  }
+
+  DERIVATIVE states {
+    rates(v)
+    m' = (minf -m)/1.5
+  }
+
+  PROCEDURE rates(v (mV)) {
+    minf = minf+1
+  }
+
+The above MOD file needs to be changed into the following one to be compatible with CoreNEURON:
+
+.. code-block:: c++
+
+  NEURON {
+    SUFFIX test
+    RANGE minf : turned minf variable to RANGE
+    THREADSAFE : added THREADSAFE keyword and moved NEURON block before any VERBATIM block
```

Comment:
I suggest ```THREADSAFE``` be the first keyword in the NEURON block. That usually does not matter and I don't recall the exact circumstances where it does matter (or even if that is still the case). It would be necessary for me to review the implementation in regard to what, if anything, can go wrong if variables are declared  prior to setting the THREADSAFE flag.

Reply:
In case there is any issue I think we should suggest adding THREADSAFE in the beginning of the NEURON block then. I edited the document based on that